### PR TITLE
LJ adagrams

### PR DESCRIPTION
**Pull request by elle-johnnie**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | From ruby I was able to carry over similar control flow and logic patterns to break the problem down.
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | Not really.
What was a challenge you were able to overcome on this assignment? | Debugging with console.log was kind annoying. I also realized late in the game that using string interpolation when calling variables resulted in less than desirable output that threw me off for many hours...but now I see what's up. 
What is your favorite thing about learning a new programming language? | All the mental hurdles makes for good mental exercise!
What is your least favorite thing about learning a new programming language? | Slamming my mental shins on all the mental hurdles. Also functional programming still feelings like a giant brick wall sitting in my path...so now I'm just waiting for that jetpack of knowledge/understanding to appear to launch me over it. 
Do you have any recommendations on how we could improve this project for the next cohort? | No immediate thoughts on this one, the built in tests made it a more clear cut process in breaking the game down to smaller functions.

### CODE REVIEW

File: src/adagrams.js [hunk 1]
```diff
@@ -1,8 +1,153 @@
...
+        return lettersInhHand
+    },
+    // ensure that only lettersInHand valid inputs
+    // create an hash object that sums each letterInHand
+    // k:letter, v:letter count
+    usesAvailableLetters(input, lettersInHand) {
+
+        let count = lettersInHand.reduce((tally, letter) => {
+            // ????? I don't know exactly how this next line works but it does :| ?????????
+            tally[letter] = (tally[letter] || 0) + 1;
```
Comment (Hamled) [inline src/adagrams.js:58]: So this line breaks down as follows:
```js
(tally[letter] || 0)
```

This is the first bit that gets evaluated because we always evaluate the _right-hand side_ of an assignment statement first, and then because the parentheses say to evaluate this bit before the addition bit. It basically says, if `tally[letter]` is empty, then "default" to 0.

It works because `tally[letter]` will be `undefined` if that letter was not added into the `tally` object by a previous iteration of this loop. And `(undefined || 0)` is equivalent to 0.

Afterwards we add 1 to it and that sum is the new value for `tally[letter]`. This code is all just a one-line version of the following:
```js
if(tally[letter] === undefined) {
  tally[letter] = 0;
}
tally[letter] += 1;
```

File: src/adagrams.js [hunk 2]
```diff
@@ -1,8 +1,153 @@
...
+            if (w.length === 10) {
+                temp.push(w);
+            }
+        }
+        // if > 1 word uses all 10 char, pick 1st in order
+        if (temp.length >= 1) {
+            max[0] = temp[0];
+        }
+        //???????? why is my IDE telling me that the variable winner is redundant????????
+        const winner = {
```
Comment (Hamled) [inline src/adagrams.js:139]: Mostly likely that warning / linter message is just because the computer sees that the only thing you do with `winner` is return it, so you could simplify the code from:
```js
const winner = {
  score: this.scoreWord(max[0]),
  word: max[0],
};

return winner;
```

to:
```js
return {
  score: this.scoreWord(max[0]),
  word: max[0],
};
```

**Comment by Hamled** [comment]
# JS Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | :white_check_mark:<br>I hope that you're feeling fewer of those mental hurdles as we've moved through the JS curriculum. And I agree we should try to provide y'all better debugging tools for the first week.
Small commits with meaningful commit messages | :white_check_mark:
**Code Requirements** |
`drawLetters` method | :white_check_mark:
Uses appropriate data structure to store the letter distribution | :white_check_mark:
All tests for `drawLetters` pass | :white_check_mark:
`usesAvailableLetters` method | :white_check_mark:
All tests for `usesAvailableLetters` pass | :white_check_mark:
`scoreWord` method | :white_check_mark:
Uses appropriate data structure to store the letter scores | :white_check_mark:
All tests for `scoreWord` pass | :white_check_mark:
`highestScoreFrom` method | :white_check_mark:
Appropriately handles edge cases for tie-breaking logic | :white_check_mark:
All tests for `highestScoreFrom` pass | :white_check_mark:
**Overall** | I like your use of lodash! While you've used it here for the sampling functionality it has built-in, lodash is primarily a general purpose functional programming library. As such, there are a lot of places where we could make use of its features to simplify our code.<br>I did an implementation of an earlier version of this project using lodash a couple of years ago. Some of the rules for the project were different so the code isn't 1-1, but you might be able to see how various pieces (like `filter`, `map`, and `reduce`) can be combined to implement most of the logic: https://github.com/Hamled/js-scrabble/blob/master/src/scrabble.js
